### DISCUSSION TITLE
Can I reference mongoose models from typegoose?

### DISCUSSION THREAD

**Discussion by JanKalkan** [typegoose/typegoose#817] [Q&A] [answered]
Hi all,

we are migrating our codebase to TS. Currently, we are using Mongoose and define quite complex models via mongoose.Schema class. We are evaluating migrating to Typegoose and would prefer to migrate incrementally, i.e. model by model and not all at once. The problem is that our models heavily ref each other and we would need to ref a mongoose model in this case. Is this possible or is there another approach we could take? My main concern is that existing code still needs to work (e.g. query population).  I already tried to speed up that migration w/ Chat-GPT but the migrated classes are not really usable.

Thanks in advance 🙏🏽

**Comment by hasezoey** [comment]
typegoose's models / classes are compatible with plain mongoose's, you will just need to explicitly define some things.

for example, you will need to be explicit about which model you are referencing, like in plain mongoose via a hard-coded string:

```ts
interface IOldSchema {
  reference: typegoose.Ref<TypegooseClass>;
}


const oldSchema = new mongoose.Schema({
  reference: {
    ref: 'TypegooseClass'
  }
});

const OldModel = mongoose.model<IOldSchema>("OldModel", oldSchema);

class TypegooseClass {
  @prop({ ref: "OldModel" })
  public reference: mongoose.PopulatedDoc<IOldSchema>;
}

const TypegooseModel = getModelForClass(TypegooseClass);
```

Note that this above is basically pseudo-code, i have not verified that it is running, but it should outline the basic method

other things you may wanna be explicit about is how the typegoose models are named, which you can read up on in [Typegoose's Name Generation](https://typegoose.github.io/typegoose/docs/guides/advanced/name-generation)

**Reply by JanKalkan** [reply]
Thx for your reply @hasezoey! Do we need the TS interface of the respective mongoose model that we want to reference from the typegoose one? Because we only have the schema definitions and no TS interfaces created.

**Reply by hasezoey** [reply, marked as answer]
> Do we need the TS interface of the respective mongoose model that we want to reference from the typegoose one?

if you previously relied on mongoose's inferred types, you can also use those, or if you didnt rely on types previously, you could also just "workaround" it by using a generic any interface or `Record<string, any>` for the time being.

example with inferred types:

```ts
const oldSchema = new mongoose.Schema({
  reference: {
    ref: 'TypegooseClass'
  }
});

const OldModel = mongoose.model("OldModel", oldSchema);
type OldSchemaType = mongoose.InferSchemaType<typeof oldSchema>;

class TypegooseClass {
  @prop({ ref: "OldModel" })
  public reference: mongoose.PopulatedDoc<OldSchemaType>;
}

const TypegooseModel = getModelForClass(TypegooseClass);
```
Note again that this is mostly pseudo-code and not directly tested

**Reply by JanKalkan** [reply]
@hasezoey Thank you 🙏🏽
